### PR TITLE
Combine the two rules: DecorrelateWhereExists and DecorrelateWhereIn

### LINKED ISSUE #6086: Code refactor, combine the two rules: DecorrelateWhereExists and DecorrelateWhereIn
### Is your feature request related to a problem or challenge?

Combine the two rules to one, the process of the two rules are almost the same and there are duplicate code.
I will try to combine the two rules to one unified rule

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #6086

# Rationale for this change
Combine the two rules into one unify rule `DecorrelatePredicateSubquery`

# What changes are included in this PR?
1.  Combine the two rules.
2. Addressed the TODO: if subquery doesn't get optimized, optimized children are lost.

# Are these changes tested?
Yes, covered by existing UTs.

# Are there any user-facing changes?

### CODE REVIEW

File: benchmarks/expected-plans/q4.txt [hunk 1]
```diff
@@ -5,13 +5,14 @@
 |               |   Projection: orders.o_orderpriority, COUNT(UInt8(1)) AS order_count                                                                                          |
```
Comment (alamb) [inline benchmarks/expected-plans/q4.txt:5]: I reviewed these plans and they see to be all the same, structurally (there are some aliases that are different)

**Review by alamb — COMMENTED** [review]
Thank you @mingmwang  -- this is looking awesome. I reviewed the  plan changes carefully and spot checked the code. 

Everything I saw looks great, but I think it would help if someone who was more familiar with this code (like @ygf11  or @jackwener , perhaps) could review it more carefully.

If they don't have time, I will try and find time to fully review this PR later this week

**Comment by jackwener** [comment]
I am going to review this PR carefully in the next two days, Thanks @mingmwang

File: datafusion/optimizer/src/decorrelate_predicate_subquery.rs [hunk 2]
```diff
@@ -54,7 +59,7 @@ impl DecorrelateWhereIn {
         predicate: &Expr,
         config: &dyn OptimizerConfig,
     ) -> Result<(Vec<SubqueryInfo>, Vec<Expr>)> {
-        let filters = split_conjunction(predicate); // TODO: disjunctions
+        let filters = split_conjunction(predicate); // TODO: add ExistenceJoin to support disjunctions
```
Comment (jackwener) [inline datafusion/optimizer/src/decorrelate_predicate_subquery.rs:62]: This PR already support disjunction of scalar/in subquery?

Comment (mingmwang) [reply datafusion/optimizer/src/decorrelate_predicate_subquery.rs:62]: No, it doesn't.  To support disjunction of predicate subqueries, need to add a specific join type  `ExistenceJoin`

Comment (jackwener) [reply datafusion/optimizer/src/decorrelate_predicate_subquery.rs:62]: Ok, BTW, there is a issue to track it `MarkJoin` #5368

**Review by jackwener — APPROVED** [review]
A great job to me

**Comment by mingmwang** [comment]
@jackwener @alamb 

Can we merge this PR, so that I can continue working on the other sub query related improvements/enhancements.

**Comment by jackwener** [comment]
> @jackwener @alamb
> 
> Can we merge this PR, so that I can continue working on the other sub query related improvements/enhancements.

I think we can push this PR to continue to do more work

**Comment by alamb** [comment]
THanks @jackwener and @mingmwang

**Comment by alamb** [comment]
I believe this caused a minor diff on main causing Ci to fail:  https://github.com/apache/arrow-datafusion/pull/6353

**Pull request merged**